components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,341 @@
+	utils.PanicCapturingGo(func() {
+		defer o.activeBackgroundWorkers.Done()
+		ticker := time.NewTicker(time.Duration(o.timeIntervalMSecs) * time.Millisecond)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Sleep until it's time to update the position again.
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			// always use the first pair until more than one pair of motors is supported in this model
+			leftMove, err := o.motors[0].left.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			rightMove, err := o.motors[0].right.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			if !leftMove && !rightMove {
+				// neither motors are moving, so the odometry values do not need to be updated
+				o.linearVelocity.Y = 0
+				o.angularVelocity.Z = 0
+				continue
+			}
+
+			// use GetInParallel to ensure the left and right motors are polled at the same time
+			positionFuncs := func() []rdkutils.FloatFunc {
+				fs := []rdkutils.FloatFunc{}
+
+				o.mu.Lock()
+				defer o.mu.Unlock()
+				// always use the first pair until more than one pair of motors is supported in this model
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].left.Position(ctx, nil) })
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].right.Position(ctx, nil) })
+
+				return fs
+			}
+
+			_, positions, err := rdkutils.GetInParallel(ctx, positionFuncs())
+			if err != nil {
+				o.logger.Error(err)
+				continue
+			}
+
+			// current position of the left and right motors in revolutions
+			if len(positions) != len(o.motors)*2 {
+				o.logger.Error("error getting both motor positions, trying again")
+				continue
+			}
+			left := positions[0]
+			right := positions[1]
+
+			// difference in the left and right motors since the last iteration, in mm
+			leftDist := (left - o.lastLeftPos) * o.wheelCircumference
+			rightDist := (right - o.lastRightPos) * o.wheelCircumference
+
+			// update lastLeftPos and lastRightPos to be the current position in mm
+			o.lastLeftPos = left
+			o.lastRightPos = right
+
+			// linear and angular distance the center point has traveled
+			centerDist := (leftDist + rightDist) / 2
+			centerAngle := (rightDist - leftDist) / o.baseWidth
+
+			// update the position and orientation values accordingly
+			o.mu.Lock()
+			o.orientation.Yaw += centerAngle
+
+			o.orientation.Yaw = math.Mod(o.orientation.Yaw, oneTurn)
+			o.orientation.Yaw = math.Mod(o.orientation.Yaw+oneTurn, oneTurn)
+
+			// calculate X and Y by using centerDist and the current orientation yaw (theta)
+			o.position.X += (centerDist * math.Cos(o.orientation.Yaw))
+			o.position.Y += (centerDist * math.Sin(o.orientation.Yaw))
+
+			// update the linear and angular velocity values using the provided time interval
+			o.linearVelocity.Y = centerDist / (o.timeIntervalMSecs / 1000)
```

Comment:
[semi-important] The position seems absolute: if you move forward, it'll increase X without changing Y, but if you turn a right angle and then move forward, it'll increase Y without changing X. However, the velocity seems relative: you only ever change the Y velocity, no matter which direction you're facing. Is that going to be confusing? I'd kinda expect the velocity to point in the same direction the robot is pointing, so that integrating the velocity over time should give you the position.

This should be super easy to do:
```go
oldPositionX = o.position.X
oldPositionY = o.position.Y

o.position.X = centerDist * math.Cos(Yaw)
o.position.X = centerDist * math.Sin(Yaw)

o.linearVelocityX = (o.position.X - oldPositionX) / (o.timeIntervalMSecs / 1000)
o.linearVelocityY = (o.position.Y - oldPositionY) / (o.timeIntervalMSecs / 1000)
```

Reply:
that's a very good point. my only thought is like if you're driving the rover straight, but say theta is like a 45 degree angle, wouldn't it be like half `linearVelocity.X` and half `linearVelocity.Y`? maybe that's what's correct but it's not really intuitive to me if feel like you'd want to just be the forward velocity basically. like no matter what direction the base is facing it's going 1 m/s linearly